src/nix/search.cc

```diff
@@ -263,11 +282,13 @@ struct CmdSearch : SourceExprCommand, MixJSON
                 throw SysError("cannot rename '%s' to '%s'", tmpFile, jsonCacheFileName);
         }
 
-        if (results.size() == 0)
-            throw Error("no results for the given search term(s)!");
+        if (!json) {
+            if (results.size() == 0)
+                throw Error("no results for the given search term(s)!");
 
-        RunPager pager;
-        for (auto el : results) std::cout << el.second << "\n";
+            RunPager pager;
+            for (auto el : results) std::cout << el.second << "\n";
```

Comment:
Don't we need to run this line when `json == true` as well?

Reply:
when using json, the user can be expected to understand what an empty result means

Reply:
After taking another look at the code, it appears the real answer is "when `json == true`, `results` is always empty".